File: samples/preview.py
"""
=============================================================================
Obsidian Neon (Dracula Syntax) - Python Theme Showcase
=============================================================================
Demonstrating:
  - Ghost/Cursive docstrings and inline comments
  - Decorators & Class definitions with Generics
  - Async/Await, Pattern Matching & Comprehensions
  - Semantic token highlighting, Constants & Numbers
"""

from __future__ import annotations
import asyncio
import re
from dataclasses import dataclass, field
from datetime import datetime, timezone
from enum import Enum, auto
from typing import (
    Any,
    Callable,
    Dict,
    Generic,
    List,
    Optional,
    Tuple,
    TypeVar,
    Union,
)

# --- Constants & Regular Expressions ---
DEFAULT_THEME: str = "Obsidian Neon (Dracula Syntax)"
PALETTE_HEX_CODES: Tuple[str, ...] = ("#191525", "#BD93F9", "#FF79C6", "#50FA7B", "#F1FA8C", "#8BE9FD")
HEX_COLOR_REGEX: re.Pattern = re.compile(r"^#([A-Fa-f0-9]{6}|[A-Fa-f0-9]{8})$")
MAX_CACHE_LIMIT: int = 1_000_000
RATIO_SCALE: float = 3.141592653589793
FEATURE_FLAGS: int = 0b1010_1100_0001


class ThemeMode(Enum):
    """Enumeration for supported UI theme variants."""
    OBSIDIAN_DEFAULT = auto()
    DRACULA_SYNTAX = auto()
    HIGH_CONTRAST = auto()


T = TypeVar("T")
R = TypeVar("R")


def timed_execution(label: str) -> Callable[[Callable[..., R]], Callable[..., R]]:
    """Decorator to measure and log execution time with high precision."""
    def decorator(func: Callable[..., R]) -> Callable[..., R]:
        async def async_wrapper(*args: Any, **kwargs: Any) -> R:
            start_ts = datetime.now(timezone.utc)
            try:
                result = await func(*args, **kwargs)
                return result
            finally:
                elapsed = (datetime.now(timezone.utc) - start_ts).total_seconds() * 1000
                print(f"[TRACE] {label} -> {func.__name__} completed in {elapsed:.2f}ms")
        return async_wrapper  # type: ignore[return-value]
    return decorator


@dataclass(slots=True, frozen=True)
class NeonColor:
    """Immutable representation of a 32-bit RGBA color token."""
    red: int
    green: int
    blue: int
    alpha: float = 1.0
    name: Optional[str] = None
    tags: List[str] = field(default_factory=list)

    @property
    def hex_code(self) -> str:
        """Converts RGB channels to upper-case Hex notation."""
        return f"#{self.red:02X}{self.green:02X}{self.blue:02X}"

    def to_css_rgba(self) -> str:
        return f"rgba({self.red}, {self.green}, {self.blue}, {self.alpha:.2f})"


class SyntaxEngine(Generic[T]):
    """Generic syntax highlighting processor for Obsidian Neon theme."""

    def __init__(self, mode: ThemeMode = ThemeMode.DRACULA_SYNTAX) -> None:
        self._mode: ThemeMode = mode
        self._color_registry: Dict[str, NeonColor] = {}
        self._is_active: bool = True
        self._bootstrap_palette()

    def _bootstrap_palette(self) -> None:
        """Populate initial Dracula Neon vibrant token palette."""
        self._color_registry = {
            "background": NeonColor(25, 21, 37, name="Obsidian Base"),
            "keyword": NeonColor(255, 121, 198, name="Hot Pink Italic"),
            "function": NeonColor(80, 250, 123, name="Neon Green Bold"),
            "type": NeonColor(139, 233, 253, name="Cyan Type"),
            "string": NeonColor(241, 250, 140, name="Lemon String"),
            "constant": NeonColor(189, 147, 249, name="Purple Constant"),
            "parameter": NeonColor(255, 184, 108, name="Orange Param"),
        }

    @timed_execution(label="AST_RENDER")
    async def render_token_tree(self, tokens: List[Dict[str, Any]]) -> List[str]:
        """Asynchronously parses AST nodes and applies theme colors."""
        rendered: List[str] = []
        for index, item in enumerate(tokens):
            token_type = item.get("type", "unknown")
            value = item.get("value", "")

            # Python 3.10+ Pattern Matching
            match token_type:
                case "keyword" | "storage":
                    style = f"\033[3;35m{value}\033[0m"
                case "function" if item.get("is_declaration", False):
                    style = f"\033[1;32m{value}\033[0m"
                case "string":
                    style = f"\033[33m'{value}'\033[0m"
                case "number" | "boolean":
                    style = f"\033[36m{value}\033[0m"
                case _:
                    style = str(value)

            rendered.append(style)
            await asyncio.sleep(0.001)

        return rendered

    def filter_palette(self, min_intensity: int = 100) -> Dict[str, NeonColor]:
        """Comprehension filtering tokens based on channel luminance."""
        return {
            key: color
            for key, color in self._color_registry.items()
            if max(color.red, color.green, color.blue) >= min_intensity
        }


# --- Entry Point Execution ---
async def main() -> None:
    engine = SyntaxEngine[str](mode=ThemeMode.DRACULA_SYNTAX)
    sample_tokens = [
        {"type": "keyword", "value": "def"},
        {"type": "function", "value": "calculate_glow", "is_declaration": True},
        {"type": "parameter", "value": "radius_px"},
        {"type": "string", "value": "vibrant_dracula_glow"},
        {"type": "number", "value": 42},
    ]

    try:
        output = await engine.render_token_tree(sample_tokens)
        print(f"Rendered {len(output)} tokens successfully.")
    except Exception as err:
        print(f"[ERROR] Failed to render theme pipeline: {err}")
        raise


if __name__ == "__main__":
    asyncio.run(main())

File: samples/preview.kt
package com.obsidian.neon.kotlin

/**
 * =============================================================================
 * Obsidian Neon (Dracula Syntax) - Kotlin Showcase
 * =============================================================================
 * Demonstrates:
 *   - KDoc cursive comments and structured annotations
 *   - Data classes, Value classes, Sealed interfaces & Smart Casts
 *   - Extension functions, Operator overloading & Infix notations
 *   - Sequence DSL pipelines, Inline & Reified Generics
 *   - Null-safety (?.), Elvis operator (?:) & When expressions
 */

// --- Value Class & Data Classes ---
@JvmInline
value class HexColor(val raw: String) {
    init {
        require(raw.startsWith("#")) { "Hex color must start with '#'" }
    }
}

data class NeonPalette(
    val name: String,
    val background: HexColor = HexColor("#191525"),
    val foreground: HexColor = HexColor("#E1E1E6"),
    val pink: HexColor = HexColor("#FF79C6"),
    val green: HexColor = HexColor("#50FA7B"),
    val cyan: HexColor = HexColor("#8BE9FD"),
    val isDraculaSyntax: Boolean = true
)

// --- Sealed Hierarchy for AST Tokens ---
sealed interface SyntaxNode {
    val line: Int
    val column: Int

    data class Keyword(val text: String, override val line: Int, override val column: Int) : SyntaxNode
    data class FunctionCall(val name: String, val argCount: Int, override val line: Int, override val column: Int) : SyntaxNode
    data class StringLiteral(val content: String, override val line: Int, override val column: Int) : SyntaxNode
    data object EndOfFile : SyntaxNode {
        override val line: Int = -1
        override val column: Int = -1
    }
}

// --- Extension Functions & Operator Overloading ---
operator fun HexColor.plus(alphaPercentage: Int): String =
    "${this.raw}${alphaPercentage.coerceIn(0, 100).toString(16).padStart(2, '0')}"

infix fun String.highlightWith(color: HexColor): String =
    "\u001B[38;2;${color.raw}m$this\u001B[0m"

// --- Inline Function with Reified Type ---
inline fun <reified T : SyntaxNode> List<SyntaxNode>.filterNodeType(): List<T> =
    this.filterIsInstance<T>()

// --- Showcase Service & Sequence Processing ---
class ObsidianKotlinEngine(val palette: NeonPalette) {

    /**
     * Emits highlighting stream using Kotlin Sequence generator.
     */
    fun tokenSequence(nodes: List<SyntaxNode>): Sequence<String> = sequence {
        for (node in nodes) {
            val formatted = when (node) {
                is SyntaxNode.Keyword ->
                    "Keyword '${node.text}' -> ${palette.pink.raw} (italic cursive)"
                is SyntaxNode.FunctionCall ->
                    "Call '${node.name}()' with ${node.argCount} args -> ${palette.green.raw} (bold)"
                is SyntaxNode.StringLiteral ->
                    "Literal \"${node.content}\" -> #F1FA8C"
                SyntaxNode.EndOfFile ->
                    "EOF reached"
            }
            yield(formatted)
        }
    }

    fun renderPreview(): Int {
        val sampleNodes = listOf(
            SyntaxNode.Keyword("suspend", 1, 1),
            SyntaxNode.Keyword("fun", 1, 9),
            SyntaxNode.FunctionCall("applyNeonGlow", 2, 1, 13),
            SyntaxNode.StringLiteral("Obsidian Neon active", 1, 28),
            SyntaxNode.EndOfFile
        )

        val keywords = sampleNodes.filterNodeType<SyntaxNode.Keyword>()
        println("Found ${keywords.size} keyword tokens in scope.")

        tokenSequence(sampleNodes).forEach { log ->
            println("[Sequence] $log")
        }

        return keywords.size
    }
}

// --- Main Runner ---
fun main() {
    println("🔮 Launching Obsidian Neon Kotlin Showcase...")
    val palette = NeonPalette("Obsidian Neon Dracula")
    val engine = ObsidianKotlinEngine(palette)

    val count = engine.renderPreview()
    println("Engine rendered successfully with $count tokens processed.")
}
